### ISSUE TITLE
Error when using #write_to_file with tempfiles

### ISSUE THREAD

**Issue by GustavoCaso** [jcupitt/ruby-vips#154]
Hi I'm currently working to add support for `ruby-vips` for the gem [image_processing](https://github.com/janko-m/image_processing)

This is the PR I'm actually working on https://github.com/janko-m/image_processing/pull/14

Currently, it supports both destructive `method_name!` and non-destructive, so the meat comes to this method: `with_ruby_vips`

```ruby
def crop(file, width, height, gravity: 'NorthWest', destination_path: nil, &block)
      with_ruby_vips(file, destination_path) do |vips_image|
        top, left = Gravity.get(vips_image, width, height, gravity)
        vips_image.crop top, left, width, height
      end
    end

def crop!(file, width, height, gravity: "NorthWest", &block)
  crop(file, width, height, gravity: gravity, destination_path: file.path, &block)
end

def with_ruby_vips(file, destination_path = nil, extension: nil)
      file_extension = extension || File.extname(file.path)
      vips_image = ::Vips::Image.new_from_file file.path
      vips_image = yield(vips_image) if block_given?
      destination_file = Utils.destination_file(destination_path, file_extension)
      vips_image.write_to_file(destination_file.path)
      destination_file
end

def destination_file(path, extension)
  if path && File.file?(path)
    File.new(path)
  else
    tempfile(extension)
  end
end

def tempfile(extension)
  Tempfile.new(["image_processing-vips", extension.to_s], binmode: true)
end
```

So I want to test both implementations, but when it gets to test the destructive one I get 
```ruby
Vips::Error: VipsJpeg: Empty input file
VipsRegion: valid clipped to nothing

/Users/gustavo.caso/dev/personal/ruby/image_processing/vendor/bundle/gems/ruby-vips-2.0.9/lib/vips/operation.rb:74:in `build'
    /Users/gustavo.caso/dev/personal/ruby/image_processing/vendor/bundle/gems/ruby-vips-2.0.9/lib/vips/operation.rb:328:in `call'
    /Users/gustavo.caso/dev/personal/ruby/image_processing/vendor/bundle/gems/ruby-vips-2.0.9/lib/vips/image.rb:411:in `write_to_file'
```

So for testing I use tempfiles to keep things cleaner, usually, the type of file that will write in the test suite would look like `"/var/folders/y7/x_h803zj2nn94ymdy73v4sdjrzn_10/T/test20180218-57452-1y7cdz3.jpg"`

I don't know if this a bug or probably a fail implementation from my side :D

**Comment by jcupitt** [comment]
Hello, you can’t use the same file for input and output, could that be the issue?

**Comment by GustavoCaso** [comment]
Well, actually that might be the problem. Is it by design of the `ruby-vips` library?

**Comment by jcupitt** [comment]
Yes, it's the way the library works. Most image processing libraries will read the input image into memory, then process it, then write to the output file. This means you can use the same file for input and output, because there is a copy of the whole thing in memory.

libvips streams the image from the input to the output in small sections. It never has the whole thing in memory at once, and all parts of the processing pipeline execute at the same time and in parallel. Because it will start writing the output before it has finished reading the input, they must be different files.

**Comment by jcupitt** [comment]
If you don't use `"sequential"` access mode, it will usually (it depends on the details of the format) decode to memory (or a temp file for large images) on open. But it will delay this decode until the first pixels are needed, which will be after it has initialized the image write. Initializing write will normally create a new empty file, and that will destroy the input before it has been read :(

**Comment by jcupitt** [comment]
The libvips test suite in pyvips avoids this problem by adding a serial number to temp file names. Perhaps that could work for you?

**Comment by GustavoCaso** [comment]
@jcupitt Thank you so much for the information I will try it and let you know how it goes. Also thank you so much for the detailed explanation.

**Comment by jcupitt** [comment]
Oh no problem. I didn't know about your PR on `image_processing`, it looks really nice!

**Comment by GustavoCaso** [comment]
Thanks @jcupitt ❤️

**Comment by janko-m** [comment]
Thank you for explaining why `libvips` doesn't have in-place processing functionality. It makes total sense, and I like that it works the way that it does.

In the `image_processing` branch I've removed processing in-place, because I realized it wasn't actually necessary; the equivalent ImageMagick functionality has led to some confusion in the past when in-place processing is used, I think people never actually want it.

So as far as we're concerned, this issue can be closed.
